visor/binaries_runner.go

```diff
@@ -0,0 +1,177 @@
+// Copyright (c) 2022 Gobalsky Labs Limited
+//
+// Use of this software is governed by the Business Source License included
+// in the LICENSE file and at https://www.mariadb.com/bsl11.
+//
+// Change Date: 18 months from the later of the date of the first publicly
+// available Distribution of this version of the repository, and 25 June 2022.
+//
+// On the date above, in accordance with the Business Source License, use
+// of this software will be governed by version 3 or later of the GNU General
+// Public License.
+
+package visor
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"path"
+	"sync"
+	"syscall"
+	"time"
+
+	"code.vegaprotocol.io/vega/logging"
+	"code.vegaprotocol.io/vega/visor/config"
+	"code.vegaprotocol.io/vega/visor/utils"
+
+	"golang.org/x/sync/errgroup"
+)
+
+type BinariesRunner struct {
+	mut         sync.RWMutex
+	running     map[string]*exec.Cmd
+	binsFolder  string
+	log         *logging.Logger
+	stopTimeout time.Duration
+}
+
+func NewBinariesRunner(log *logging.Logger, binsFolder string, stopTimeout time.Duration) *BinariesRunner {
+	return &BinariesRunner{
+		binsFolder:  binsFolder,
+		running:     map[string]*exec.Cmd{},
```

Comment:
not sure if it matters, but would it make sense to keep them in order implied in the slice passed to run()? as is, the order you signal them is arbitrary.

Reply:
IMHO it should not matter. Because we `sigterm` for graceful shutdown.